### PR TITLE
[RELAY] Move Layout to tvm Node system

### PR DESCRIPTION
**Pull request by merrymercy**
modify the data structure `Layout` in `src/relay/op/nn/layout.h` to a tvm Node

@tqchen

### CODE REVIEW

**Comment by tqchen** [comment]
I think we can still keep the data structure for now in private header, as we have not decided on we should proceed on enabling it. BTW, it is fine to have layout as a field of node class, we just may not be able to access it from python side(need to use string access)

**Comment by merrymercy** [comment]
This is required as I need users to register infer_layout function for operators.
This signature must be in public header, so the layout node should also be in public header.

```c++
using FInferCorrectLayout = runtime::TypedPackedFunc<
  Array<Array<Layout> >(const Attrs& attrs,
                        const Array<Layout>& inputs,
                        const Array<Layout>& last_inputs)>;
```

File: include/tvm/relay/layout.h [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+  static bool is_superdim(LayoutDim dim) {
+    return dim >= 'A' && dim <= 'Z';
+  }
+
+  /*!
+   * \brief Check whether a given dimension is a sub-dimension.
+   * \param dim input dimension
+   * \return Whether a given dimension is a sub-dimension.
+   */
+  static bool is_subdim(LayoutDim dim) {
```
Comment (jroesch) [inline include/tvm/relay/layout.h]: Can we have these methods stylistically match the rest of the code base i.e `IsSubdim`

**Comment by yzhliu** [comment]
@merrymercy I'm implementing in this way https://discuss.tvm.ai/t/datalayout-structure/80 , basically providing a bijection approach. Want to hear your opinion about that.

**Comment by tqchen** [comment]
close this for now as Layout might be updated as @yzhliu mentioned, we can hide the current layout system behind a Node

**Comment by merrymercy** [comment]
I can move the header file to a private source directory. But I do need this data structure in tvm node system.

So I can
1. Use this pr
2. Or register another node such as LayoutInfo, which only contains one filed Layout. The drawback is that we will introducing confusing names.

I think we can merge this pr since this pr is very clean.
This pr only changes the header file while does not change any usage.
Besides, this pr does not bring any overhead to latter refactor. On the contray, it will benefit latter refactor.

File: include/tvm/relay/layout.h (hunk 1 above)
Comment (merrymercy) [reply include/tvm/relay/layout.h]: addressed

**Comment by tqchen** [comment]
@merrymercy is it ready to be merged? @jroesch please https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

**Comment by merrymercy** [comment]
ready for merge!

**Pull request merged**